### PR TITLE
MNT dont pervert BUILD if state_setter is specified

### PR DESCRIPTION
**Pull request by pierreglaser**
@ogrisel @pitrou @ncoghlan here is another way of saving an object with a custom `state_setter`. 
This is more compact than the other option, and has the advantage of being fully backward compatible, but it involves some Unpickler stack manipulation:

stack before the state handling part of `save_reduce`
(1) `[..., obj]`

if `state_setter is not None`, we write state_setter and its args, with the resulting stack

(2) `[..., obj, state_setter, (obj, state)]`

after `state_setter` is called, the stack is 
(3) `[..., obj, state_setter(obj, state)]`

and we finally have to discard the top of stack (as `state_setter`'s purpose is to do in-place modification of `obj`), by writing a POP opcode.

WDYT?

### CODE REVIEW

**Comment by pitrou** [comment]
If you manage to write a detailed comment that explains it then why not ;-)

Feel free to post a pickle disassembly of an example here :-)

File: Lib/test/pickletester.py [hunk 1]
```diff
@@ -3006,7 +3006,6 @@ def setstate_bbb(obj, state):
     """
     obj.a = 'foo'
 
-
```
Comment (ogrisel) [inline Lib/test/pickletester.py]: spurious unrelated line deletion

**Review by ogrisel — APPROVED** [review]
LGTM besides the following:

**Comment by pierreglaser** [comment]
Here are two disassembled pickles of an instance. The first pickle uses `load_build` to update the instance at unpickling time, whereas the second uses a custom state-setter.
If this looks fine to you, I'll merge this PR, and write the pure python `pickle` equivalent. 

```
    0: \x80 PROTO      4
    2: \x95 FRAME      25
   11: \x8c SHORT_BINUNICODE '__main__'
   21: \x8c SHORT_BINUNICODE 'MyClass'
   30: \x93 STACK_GLOBAL
   31: )    EMPTY_TUPLE
   32: R    REDUCE
   33: }    EMPTY_DICT
   34: b    BUILD
   35: .    STOP
highest protocol among opcodes = 4
```

second pickle (with a custom state-setter)
```
    0: \x80 PROTO      4
    2: \x95 FRAME      56
   11: \x8c SHORT_BINUNICODE '__main__'
   21: \x94 MEMOIZE    (as 0)
   22: \x8c SHORT_BINUNICODE 'MyClass'
   31: \x93 STACK_GLOBAL
   32: )    EMPTY_TUPLE
   33: R    REDUCE
   34: \x94 MEMOIZE    (as 1)
   35: h    BINGET     0
   37: \x8c SHORT_BINUNICODE 'myclass_state_setter'
   59: \x93 STACK_GLOBAL
   60: h    BINGET     1
   62: }    EMPTY_DICT
   63: \x86 TUPLE2
   64: R    REDUCE
   65: 0    POP
   66: .    STOP
highest protocol among opcodes = 4
```

Reproducer: 
```python
import io
import pickletools
import _pickle


class MyClass:
    pass


def myclass_state_setter(obj, state):
    obj.my_attr = 10


def myclass_reduce(obj):
    # remove the last item of the tuple (myclass_state_setter) to use
    # load_build instead of a custom state-setter
    return MyClass, (), {}, None, None, myclass_state_setter


obj = MyClass()

f = io.BytesIO()
p = _pickle.Pickler(f)

p.dispatch_table = {MyClass: myclass_reduce}

p.dump(obj)
pickletools.dis(pickletools.optimize(f.getvalue()))
```
